File: src/number_square_cpp/src/square_subscriber.cpp
#include "rclcpp/rclcpp.hpp" // Importing the ros2 c++ client library
#include "std_msgs/msg/int32.hpp" // Importing the integer message type

//Creating a node class SquareSubscriber
class SquareSubscriber : public rclcpp::Node {
public:
    //Constructing the node named square_subscriber
    SquareSubscriber() : Node("square_subscriber") {
        
        //Setting up a QoS profile
        auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_default))
            .reliable() //Ensures message is recieved retransmits if not
            .durability_volatile() //Future nodes can't access already published data
            .keep_last(10); // Stores 10 messages

        //Creating a subscription on topic 'number' message type Int32 and a queue size of 10
        subscription_ = this -> create_subscription<std_msgs::msg::Int32>(
            "number", qos,
            //Lambda function to capture the message and log its square
            [this](const std_msgs::msg::Int32::SharedPtr msg) {
                int num = msg -> data;
                int square = num * num;
                RCLCPP_INFO(this->get_logger(), "Recieved: %d, Square: %d", num, square);
            }
        );
    }

private:
    rclcpp::Subscription<std_msgs::msg::Int32>::SharedPtr subscription_;
};

int main(int argc, char *argv[]) {
    //Initializing the ros2 library
    rclcpp::init(argc, argv);
    //Spin makes the node SquareSubscriber stay alive
    //Using make_shared helps manage the lifecycle of the node effeciently
    rclcpp::spin(std::make_shared<SquareSubscriber>());
    //Shutting the node down
    rclcpp::shutdown();
    return 0;
}

File: src/number_square_py/number_square_py/square_subscriber.py
import rclpy #Importing the ros client library for python
from rclpy.node import Node # Importing Node class from rclpy
from std_msgs.msg import Int32 #Importing the standard message type 32-bit integer
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy

#Creaitng a node named SquareSubscriber 
class SquareSubscriber(Node):

    def __init__(self):
        # Creates a node named square_subscriber
        super().__init__('square_subscriber')
        qos_profile = QoSProfile(
            reliability = ReliabilityPolicy.RELIABLE, 
            durability = DurabilityPolicy.VOLATILE,
            depth = 10
        )
        """
        Creating a subscription of
        Topic : number
        message type : Int32
        queue size : 10"""
        self.subscription = self.create_subscription(
            Int32,
            'number',
            self.listener_callback,
            qos_profile
        )
        self.subscription # Prevent unused variable warning

    # Function to square the published data and log that info
    def listener_callback(self, msg):
        square = msg.data ** 2
        self.get_logger().info(f"Recieved: {msg.data}, Square : {square}")

def main(args = None):
    #Initializing the ros library
    rclpy.init(args=args)
    #Creating the subscriber
    square_subscriber = SquareSubscriber()
    #Spin the node to keep it alive
    rclpy.spin(square_subscriber)

    #Clean up resources once node is closed
    square_subscriber.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()
